src/pubsub/ua_pubsub_keystorage.h

```diff
@@ -0,0 +1,403 @@
+
+    /**
+     * id used to register the callback to retrieve the keys related to this security
+     * group
+     */
+    UA_UInt64 callBackId;
+
+    /**
+     * Pointer to the key storage list
+     */
+    LIST_ENTRY(UA_PubSubKeyStorage) keyStorageList;
+
+} UA_PubSubKeyStorage;
+
+/**
+ * @brief Allocates memory for a UA_PubSubKeyStorage and returns a pointer of the
+ * PubSubKeyStorage. The memory must be freed by calling UA_PubSubKeyStorage_delete or
+ * UA_PubSubKeyStorage_removeKeyStorage
+ *
+ * @return UA_PubSubKeyStorage*
+ */
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_new(void);
+
+/**
+ * @brief Find the Keystorage from the Server KeyStorageList and returns the pointer to
+ * the keystorage
+ *
+ * @param server holds the keystoragelist
+ * @param securityGroupId of the keystorage to be found
+ * @param keyStorage is the pointer to returned keystorage, if not found null is returned
+ * @return Pointer to the keystorage on success, null pointer on failure
+ */
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_getKeyStorageWithSecurityGroupId(UA_Server *server,
+                                                     UA_String securityGroupId);
+
+/**
+ * @brief retreives the security policy pointer from the PubSub configuration by
+ * SecurityPolicyUri
+ *
+ * @param server the server object
+ * @param securityPolicyUri the URI of the security policy
+ * @param policy the pointer to the security policy
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_getPubSubSecurityPolicyByUri(UA_Server *server,
+                                                 const UA_String *securityPolicyUri,
+                                                 UA_PubSubSecurityPolicy **policy);
+
+/**
+ * @brief Deletes all member of a key storage.
+ *
+ * @param server server from which to remove the key storage
+ * @param keyStorage keyStorage to delete
+ */
+void
+UA_PubSubKeyStorage_deleteMembers(UA_Server *server, UA_PubSubKeyStorage *keyStorage);
+
+/**
+ * @brief Deletes the keystorage from the server and its members
+ *
+ * @param server where the keystorage is created
+ * @param keyStorage pointer to the keystorage
+ */
+void
+UA_PubSubKeyStorage_delete(UA_Server *server, UA_PubSubKeyStorage *keyStorage);
+
+/**
+ * @brief Initializes an empty Keystorage for the SecurityGroupId and add it to the Server
+ * KeyStorageList
+ *
+ * @param server the server object
+ * @param securityGroupId the identifier of the SecurityGroup
+ * @param securityPolicyUri the security policy assocaited with the security algorithm
+ * @param maxPastKeyCount maximum number of past keys a keystorage is allowed to store
+ * @param maxFutureKeyCount maximum number of future keys a keystorage is allowed to store
+ * @param keyStorage pointer to the keystorage to be initialized
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_initialize(UA_Server *server, const UA_String *securityGroupId,
```

Comment:
Shorten to _init to stay within the nomenclature of the SDK.

Reply:
done